safety-oracle.tex

```diff
@@ -0,0 +1,1552 @@
+\section{Safety Oracles}
+Here, we present a minimum-viable safety oracle, and prove that it is sufficient to detect safety in some situations.
+
+\subsection{Preliminary Definitions}
+
+\begin{defn}[Later]
+$$
+Later:M \times M \to \{True, False\}
+$$
+$$
+Later(m_1, m_2) :\Leftrightarrow m_2 \in Justification(m_1)
+$$
+We may use the infix operator:
+$$
+m_1 < m_2 :\Leftrightarrow m_2 \in Justification(m_1)
+$$
+% what's a better infix operator?
+\end{defn}
+
+\begin{defn}[From sender]
+$$
+From\_Sender: \mathcal{V} \times \mathcal{P}(M) \to \mathcal{P}(M)
+$$
+$$
+From\_Sender(v, \sigma) = \{m \in \sigma : Sender(m) = v\}
+$$
+\end{defn}
+
+\begin{defn}[Later messages from sender]
+$$
+Later\_From: M \times \mathcal{V} \times \mathcal{P}(M) \to \mathcal{P}(M)
+$$
+$$
+Later\_From(m, v, \sigma) = \{m' \in \sigma : Sender(m') = v \land Later(m', m)\}
+$$
+\end{defn}
+
+\begin{defn}[Latest messages]
+$$
+L_M:\Sigma \to (\mathcal{V} \to \mathcal{P}(M))
+$$
+$$
+L_M(\sigma)(v) = \{m \in \sigma : Sender(m) = v ~\land~ Later\_From(m,v,\sigma) = \emptyset\}
+$$
+\end{defn}
+
+
+\begin{defn}[Latest messages from non-Equivocating validators]
+$$
+L^H_M:\Sigma \to (\mathcal{V} \to \mathcal{P}(M))
+$$
+\[ L^H_M(\sigma)(v) = \left\{
+\begin{array}{ll}
+      \emptyset& \text{ for } v \in E(\sigma) \\
+      L_M(\sigma)(v)& \text{ otherwise }
+\end{array}
+\right. \]
+\end{defn}
+
+Note that the map returned by this function has values $L^H_M(\sigma)(v) = \emptyset$ for any validators who are equivocating in $\sigma$ or who don't have any messages in $\sigma$.
+
+\begin{defn}[Latest Estimates from non-Equivocating validators]
+$$
+L^H_E:\Sigma \to (\mathcal{V} \to \mathcal{P}(\mathcal{C}))
+$$
+$$
+L^H_E(\sigma)(v) = \{Estimate(m) : m \in L^H_M(\sigma)(v)\}
+$$
+\end{defn}
+
+As above, $L^H_E(\sigma)(v) = \emptyset$ for validators $v$ who are equivocating or missing in $\sigma$.
+
+
+\begin{defn}[Latest justifications from non-Equivocating validators]
+$$
+L^H_J:\Sigma \to (\mathcal{V} \to \mathcal{P}(\Sigma))
+$$
+$$
+L^H_J(\sigma)(v) = \{Justification(m) : m \in L^H_M(\sigma)(v)\}
+$$
+\end{defn}
+
+
+The validators whose latest estimates agree with some proposition $p$ in some protocol state $\sigma$ is denoted formally as follows
+\begin{defn}[Validators agreeing with $p$ in $\sigma$]
+$$
+Agreeing: P_{\mathcal{C}} \times \Sigma \to \mathcal{P}(\mathcal{V})
+$$
+$$
+Agreeing(p, \sigma) :\Leftrightarrow \{v \in \mathcal{V} : \forall c \in L^H_E(\sigma)(v), p(c)\}
```

Comment:
In HOL, `∀ a ∈ ∅. P` is True. Therefore, `v` s.t. `L^H_E(\sigma)(v)` is empty is included in `agreeing`.
I think it should be re-defined to remove validators which are not observed or equivocating like this (Isabelle code):
```
fun observed_non_equivocating_validators :: "state ⇒ validator set"
  where
    "observed_non_equivocating_validators σ = observed σ - equivocating_validators σ"

fun  agreeing_validators :: "(consensus_value_property * state) ⇒ validator set"
  where
    "agreeing_validators (p, σ) = {v ∈ observed_non_equivocating_validators σ. ∀ c ∈ latest_estimates_from_non_equivocating_validators σ v. p c}"
```